### PR TITLE
#429 Implement cache for QueryAccount

### PR DESCRIPTION
**Pull request by vakond**
Added method `cache` to ensure caching of an account's data to prevent inconsistencies during iterative execution and to improve overall performance.

Note: it will require new deploy of the wrapper contract QueryAccount.sol after releasing on a real network.

Integration tests are here: https://github.com/neonlabsorg/proxy-model.py/pull/398

Typical usage from a Solidity function in a contract which inherits "standard" QueryAccount contract can look as follows:
```
        ok = super.cache(solana_account, 0, 64);
        (ok, owner_account) = super.owner(solana_account);
        (ok, data_len) = super.length(solana_account); // can be >64 here
        (ok, data) = super.data(solana_account, 10, 16);
```

### CODE REVIEW

File: evm_loader/program/src/precompile_contracts.rs [hunk 1]
```diff
@@ -1,26 +1,31 @@
...
+const SYSTEM_ACCOUNT_QUERY: H160 =            H160([0xff, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0x02]);
+const SYSTEM_ACCOUNT_ECRECOVER: H160 =        H160([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0x01]);
+const SYSTEM_ACCOUNT_SHA_256: H160 =          H160([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0x02]);
+const SYSTEM_ACCOUNT_RIPEMD160: H160 =        H160([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0x03]);
+const SYSTEM_ACCOUNT_DATACOPY: H160 =         H160([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0x04]);
+const SYSTEM_ACCOUNT_BIGMODEXP: H160 =        H160([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0x05]);
+const SYSTEM_ACCOUNT_BN256_ADD: H160 =        H160([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0x06]);
+const SYSTEM_ACCOUNT_BN256_SCALAR_MUL: H160 = H160([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0x07]);
+const SYSTEM_ACCOUNT_BN256_PAIRING: H160 =    H160([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0x08]);
+const SYSTEM_ACCOUNT_BLAKE2F: H160 =          H160([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0x09]);
```
Comment (anton-lisanin) [inline evm_loader/program/src/precompile_contracts.rs:28]: Please remove unrelated line end changes from the pull request

File: evm_loader/program/src/precompile_contracts.rs [hunk 2]
```diff
@@ -301,22 +308,36 @@ const QUERY_ACCOUNT_METHOD_DATA_ID: &[u8; 4] = &[0x43, 0xca, 0x51, 0x61];
...
     let (account_address, rest) = rest.split_at(32);
     let account_address = Pubkey::new(account_address);
 
     match method_id {
+        QUERY_ACCOUNT_METHOD_CACHE_ID => {
+            // Note: abi.encodeWithSignature makes parameters padded to 32 bytes
+            let (offset, rest) = rest.split_at(32);
+            let (length, _) = rest.split_at(32);
+            let offset = U256::from_big_endian_fast(offset).as_usize();
+            let length = U256::from_big_endian_fast(length).as_usize();
```
Comment (anton-lisanin) [inline evm_loader/program/src/precompile_contracts.rs:329]: Check how arguments are parsed in `fn erc20_wrapper`.
Having the same things implemented differently in different places makes code harder to read.

File: evm_loader/program/src/precompile_contracts.rs [hunk 3]
```diff
@@ -382,13 +399,16 @@ pub fn query_account<'a, B: AccountStorage>(
             let offset = U256::from_big_endian_fast(offset).as_usize();
             let length = U256::from_big_endian_fast(length).as_usize();
             debug_print!("query_account get data {} {} {}", account_address, offset, length);
-            let data = state.query_solana_account_data(account_address, offset, length);
-            if let Some(data) = data {
-                debug_print!("query_account data result: {:?}", data);
-                return Capture::Exit((ExitReason::Succeed(evm::ExitSucceed::Returned), data));
+            match state.query_solana_account().data(&account_address, offset, length) {
+                Ok(data) => {
+                    debug_print!("query_account data result: {:?}", data);
```
Comment (anton-lisanin) [inline evm_loader/program/src/precompile_contracts.rs:404]: `data` size can be up to 8kb.
Printing it to log is unreasonable.

File: evm_loader/program/src/precompile_contracts.rs [hunk 4]
```diff
@@ -382,13 +399,16 @@ pub fn query_account<'a, B: AccountStorage>(
...
+                    Capture::Exit((ExitReason::Succeed(evm::ExitSucceed::Returned), data))
+                },
+                Err(err) => {
+                    let revert_message = format!("QueryAccount.data failed: {:?}", err).as_bytes().to_vec();
```
Comment (anton-lisanin) [inline evm_loader/program/src/precompile_contracts.rs:408]: Please use `{}` as formatter. And `enum Error` should derive `Display`

File: evm_loader/program/src/query.rs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+    /// Checks if account got data. Dataless accounts make no sense in the cache.
+    pub const fn has_data(&self) -> bool {
+        self.data.is_some()
+    }
+}
+
+/// Creates vector from a slice checking the range validity.
+fn clone_chunk(data: &[u8], offset: usize, length: usize) -> Option<Vec<u8>> {
+    if offset >= data.len() || offset + length > data.len() {
```
Comment (anton-lisanin) [inline evm_loader/program/src/query.rs:128]: What will happen if `offset + length` overflow?

File: evm_loader/program/src/executor_state.rs [hunk 6]
```diff
@@ -719,6 +722,7 @@ impl ExecutorSubstate {
 pub struct ExecutorState<'a, B: AccountStorage> {
     backend: &'a B,
     substate: Box<ExecutorSubstate>,
+    query_account_cache: query::AccountCache,
```
Comment (anton-lisanin) [inline evm_loader/program/src/executor_state.rs:725]: `query_account_cache` is not saved in a storage account and will reset between iterations.
In order to automatically save it, cache should be part of `substate`.

**Review by anton-lisanin — CHANGES_REQUESTED** [review]

**Comment by anton-lisanin** [comment]
@vakond Please review code you are including into commit before committing. And do not stage unrelated changes.

File: evm_loader/program/src/precompile_contracts.rs (hunk 1 above)
Comment (vakond) [reply evm_loader/program/src/precompile_contracts.rs:28]: Done.

File: evm_loader/program/src/precompile_contracts.rs (hunk 2 above)
Comment (vakond) [reply evm_loader/program/src/precompile_contracts.rs:329]: Done.

File: evm_loader/program/src/precompile_contracts.rs (hunk 3 above)
Comment (vakond) [reply evm_loader/program/src/precompile_contracts.rs:404]: Right. Log only size of data.

File: evm_loader/program/src/precompile_contracts.rs (hunk 4 above)
Comment (vakond) [reply evm_loader/program/src/precompile_contracts.rs:408]: Done.

File: evm_loader/program/src/query.rs (hunk 5 above)
Comment (vakond) [reply evm_loader/program/src/query.rs:128]: Check overflowing. Used saturating_add which seems enough here to check validity later.

File: evm_loader/program/src/executor_state.rs (hunk 6 above)
Comment (vakond) [reply evm_loader/program/src/executor_state.rs:725]: Moved the cache to the substate. And reworked Solidity part to inherit QueryAccount contract.

File: evm_loader/program/src/query.rs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        self.cache.get(address).map(|v| v.executable)
+    }
+
+    /// Returns rent epoch of an account if found.
+    pub fn rent_epoch(&self, address: &Pubkey) -> Option<Epoch> {
+        self.cache.get(address).map(|v| v.rent_epoch)
+    }
+
+    /// Returns chunk of data of an account if found and correct range.
+    pub fn data(&self, address: &Pubkey, offset: usize, length: usize) -> Result<Vec<u8>> {
```
Comment (anton-lisanin) [inline evm_loader/program/src/query.rs:76]: You are returning a `Result` from `.data()`, but you are using an `Option` in all other methods.
Please chose one of these variants and use it everywhere for consistency

File: evm_loader/program/src/query.rs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        debug_print!("new AccountCache");
+        Self {
+            cache: BTreeMap::new(),
+        }
+    }
+
+    /// Inserts or replaces entry into the cache.
+    pub fn put(&mut self, address: Pubkey, value: Value) {
+        self.cache.remove(&address);
+        self.cache.insert(address, value);
```
Comment (anton-lisanin) [inline evm_loader/program/src/query.rs:47]: `BTreeMap::insert` will overwrite an existing value.
`.remove()` is unnecessary.

Comment (vakond) [reply evm_loader/program/src/query.rs:47]: Indeed. Deleted the remove call.

File: evm_loader/program/src/query.rs (hunk 7 above)
Comment (vakond) [reply evm_loader/program/src/query.rs:77]: Very good point, thanks. Refactored accordingly.

**Review by anton-lisanin — APPROVED** [review]
